File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/ruby/find_shortest_string_recursive.rb
def find_shortest_string_recursive(arr)
  return arr.first if arr.length === 1
  nextString = find_shortest_string_recursive(arr[1..-1])
  arr.first.length <= nextString.length ? arr.first : nextString

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string_recursive(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string_recursive(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string_recursive(['flower', 'juniper', 'lily', 'dandelion'])
  
  puts

  puts "Expecting: 'no'"
  puts "=>", find_shortest_string_recursive(['nah', 'nope', 'no', 'nosiree'])
  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/javascript/find_shortest_string_recursive.js
function findShortestStringRecursive(arr) {
  if (arr.length === 1) {
    return arr[0]
  }
  let nextString = findShortestStringRecursive(arr.slice(1))
  return arr[0].length <= nextString.length ? arr[0] : nextString
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestStringRecursive(['aaa', 'a', 'bb', 'ccc']));

  console.log("");

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestStringRecursive(['cat', 'hi', 'dog', 'an']));

  console.log("");

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestStringRecursive(['flower', 'juniper', 'lily', 'dandelion']));

  console.log("");

  console.log("Expecting: 'no'");
  console.log("=>", findShortestStringRecursive(['nah', 'nope', 'no', 'nosiree']));
}

module.exports = findShortestStringRecursive;

// Please add your pseudocode to this file
// And a written explanation of your solution
